File: src/Scheduler.cs
using System;

namespace LumaFlux
{
    /// <summary>Ce que la planification impose a un instant donne.</summary>
    public struct ScheduleResult
    {
        public int Kelvin;
        public double Brightness;
        public bool BrightnessApplies;
        public DayPhase Phase;
        public double Sunrise, Sunset;
        public bool SunValid;
    }

    /// <summary>
    /// Traduit l'heure qu'il est en consigne de couleur et, si l'utilisateur le
    /// souhaite, de luminosite.
    ///
    /// f.lux ne fait varier que la couleur. Faire aussi descendre la luminosite le soir
    /// est ce qui manque le plus quand on utilise les deux applications d'origine
    /// cote a cote : il fallait bouger deux reglages a la main.
    /// </summary>
    public static class Scheduler
    {
        public static ScheduleResult Evaluate(Settings s, DateTime now)
        {
            ScheduleResult r = new ScheduleResult();
            r.Kelvin = s.Current.Kelvin;
            r.Brightness = s.Current.Brightness;
            r.BrightnessApplies = false;
            r.Phase = DayPhase.Day;

            Preset preset = FindPreset(s.PresetName);

            double sunrise, sunset;
            r.SunValid = SolarClock.SunTimes(now, s.Latitude, s.Longitude, out sunrise, out sunset);

            if (s.Schedule == ScheduleMode.FixedHours)
            {
                sunrise = s.FixedDayHour;
                sunset = s.FixedNightHour;
                r.SunValid = true;
            }
            else if (!r.SunValid)
            {
                // Nuit ou jour polaire : on retombe sur des horaires raisonnables.
                sunrise = 7.0;
                sunset = 19.0;
            }

            r.Sunrise = sunrise;
            r.Sunset = sunset;

            if (s.Schedule == ScheduleMode.Manual) return r;

            double fade = Math.Max(5, s.TransitionMinutes) / 60.0;
            double h = now.Hour + now.Minute / 60.0 + now.Second / 3600.0;

            DayPhase phase;
            r.Kelvin = Interpolate(preset, h, sunrise, sunset, s.BedtimeHour, fade, out phase);
            r.Phase = phase;

            if (s.ScheduleAffectsBrightness)
            {
                r.BrightnessApplies = true;
                r.Brightness = InterpolateBrightness(s, h, sunrise, sunset, fade, phase);
            }

            return r;
        }

        private static int Interpolate(Preset preset, double h, double sunrise, double sunset,
                                       double bedtime, double fade, out DayPhase phase)
        {
            if (IsBetweenWrapped(h, bedtime, sunrise))
            {
                phase = DayPhase.Late;
                return preset.Late;
            }

            if (h >= sunset - fade && h < sunset + fade)
            {
                phase = DayPhase.Sunset;
                return Lerp(preset.Day, preset.Night, (h - (sunset - fade)) / (2 * fade));
            }

            if (h >= sunrise && h < sunrise + fade)
            {
                phase = DayPhase.Sunset;
                return Lerp(preset.Night, preset.Day, (h - sunrise) / fade);
            }

            if (h >= sunset + fade || h < sunrise)
            {
                phase = DayPhase.Night;
                return preset.Night;
            }

            phase = DayPhase.Day;
            return preset.Day;
        }

        private static double InterpolateBrightness(Settings s, double h, double sunrise,
                                                    double sunset, double fade, DayPhase phase)
        {
            double day = s.DayBrightness, night = s.NightBrightness;

            if (phase == DayPhase.Day) return day;
            if (phase == DayPhase.Night || phase == DayPhase.Late) return night;

            // En transition, on suit la meme rampe que la couleur.
            if (h >= sunset - fade && h < sunset + fade)
                return day + (night - day) * ((h - (sunset - fade)) / (2 * fade));
            if (h >= sunrise && h < sunrise + fade)
                return night + (day - night) * ((h - sunrise) / fade);

            return day;
        }

        private static int Lerp(int a, int b, double p)
        {
            if (p < 0) p = 0;
            if (p > 1) p = 1;
            return (int)Math.Round(a + (b - a) * p);
        }

        private static bool IsBetweenWrapped(double h, double from, double to)
        {
            if (from <= to) return h >= from && h < to;
            return h >= from || h < to;
        }

        public static Preset FindPreset(string name)
        {
            foreach (Preset p in Preset.BuiltIn())
                if (p.Name == name) return p;
            return Preset.BuiltIn()[0];
        }

        /// <summary>Resume lisible de l'etat de la planification, pour l'interface.</summary>
        public static string Describe(Settings s, ScheduleResult r)
        {
            if (s.Schedule == ScheduleMode.Manual)
                return "Mode manuel : les curseurs commandent seuls.";

            string times = s.Schedule == ScheduleMode.FixedHours
                ? string.Format("Jour {0} - nuit {1}", SolarClock.FormatHour(r.Sunrise), SolarClock.FormatHour(r.Sunset))
                : string.Format("Lever {0} - coucher {1}", SolarClock.FormatHour(r.Sunrise), SolarClock.FormatHour(r.Sunset));

            return string.Format("{0} - actuellement : {1} ({2} K)",
                times, SolarClock.DescribePhase(r.Phase), r.Kelvin);
        }
    }
}

File: mac/Sources/OpusScreen/Engine/Scheduler.swift
import Foundation

/// Ce que la planification impose a un instant donne.
public struct ScheduleResult {
    public var kelvin: Int = 6500
    public var brightness: Double = 100
    public var brightnessApplies = false
    public var phase: DayPhase = .day
    public var sunrise: Double = 7
    public var sunset: Double = 19
    public var sunValid = false
    public init() {}
}

/// Traduit l'heure qu'il est en consigne de couleur et, si l'utilisateur le
/// souhaite, de luminosite.
///
/// f.lux ne fait varier que la couleur. Faire aussi descendre la luminosite le soir
/// est ce qui manque le plus quand on utilise les deux applications d'origine cote
/// a cote : il fallait bouger deux reglages a la main.
public enum Scheduler {

    public static func evaluate(_ s: Settings, _ now: Date) -> ScheduleResult {
        var r = ScheduleResult()
        r.kelvin = s.current.kelvin
        r.brightness = s.current.brightness
        r.brightnessApplies = false
        r.phase = .day

        let preset = findPreset(s.presetName)

        var sunrise = 7.0, sunset = 19.0
        if let t = SolarClock.sunTimes(now, s.latitude, s.longitude) {
            sunrise = t.sunrise; sunset = t.sunset; r.sunValid = true
        } else {
            r.sunValid = false
        }

        if s.schedule == .fixedHours {
            sunrise = s.fixedDayHour
            sunset = s.fixedNightHour
            r.sunValid = true
        } else if !r.sunValid {
            // Nuit ou jour polaire : on retombe sur des horaires raisonnables.
            sunrise = 7.0
            sunset = 19.0
        }

        r.sunrise = sunrise
        r.sunset = sunset

        if s.schedule == .manual { return r }

        let fade = max(5.0, Double(s.transitionMinutes)) / 60.0
        let h = SolarClock.hourOfDay(now)

        let interpolated = interpolate(preset, h, sunrise, sunset, s.bedtimeHour, fade)
        r.kelvin = interpolated.kelvin
        r.phase = interpolated.phase

        if s.scheduleAffectsBrightness {
            r.brightnessApplies = true
            r.brightness = interpolateBrightness(s, h, sunrise, sunset, fade, r.phase)
        }

        return r
    }

    private static func interpolate(_ preset: Preset, _ h: Double, _ sunrise: Double,
                                    _ sunset: Double, _ bedtime: Double, _ fade: Double)
        -> (kelvin: Int, phase: DayPhase) {

        if SolarClock.isBetweenWrapped(h, bedtime, sunrise) { return (preset.late, .late) }

        if h >= sunset - fade && h < sunset + fade {
            return (SolarClock.lerp(preset.day, preset.night, (h - (sunset - fade)) / (2 * fade)), .sunset)
        }

        if h >= sunrise && h < sunrise + fade {
            return (SolarClock.lerp(preset.night, preset.day, (h - sunrise) / fade), .sunset)
        }

        if h >= sunset + fade || h < sunrise { return (preset.night, .night) }

        return (preset.day, .day)
    }

    private static func interpolateBrightness(_ s: Settings, _ h: Double, _ sunrise: Double,
                                              _ sunset: Double, _ fade: Double, _ phase: DayPhase) -> Double {
        let day = s.dayBrightness, night = s.nightBrightness

        if phase == .day { return day }
        if phase == .night || phase == .late { return night }

        // En transition, on suit la meme rampe que la couleur.
        if h >= sunset - fade && h < sunset + fade {
            return day + (night - day) * ((h - (sunset - fade)) / (2 * fade))
        }
        if h >= sunrise && h < sunrise + fade {
            return night + (day - night) * ((h - sunrise) / fade)
        }

        return day
    }

    public static func findPreset(_ name: String) -> Preset {
        for p in Preset.builtIn() where p.name == name { return p }
        return Preset.builtIn()[0]
    }

    /// Resume lisible de l'etat de la planification, pour l'interface.
    public static func describe(_ s: Settings, _ r: ScheduleResult) -> String {
        if s.schedule == .manual { return "Mode manuel : les curseurs commandent seuls." }

        let times = s.schedule == .fixedHours
            ? "Jour \(SolarClock.formatHour(r.sunrise)) - nuit \(SolarClock.formatHour(r.sunset))"
            : "Lever \(SolarClock.formatHour(r.sunrise)) - coucher \(SolarClock.formatHour(r.sunset))"

        return "\(times) - actuellement : \(SolarClock.describePhase(r.phase)) (\(r.kelvin) K)"
    }
}
